### ISSUE TITLE
GlobalStatefulCheck anntation discussion

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#5140]
small part from #4883 , from discussion of #4898

`
  The check methods and fields should be thread safe, because they may be accessed from others
 threads at the same time.
`

Checks that are not thread safe - should be executed as it was before ..... without any thread concerns. there should be no demands for thread safety.

**Comment by romani** [comment]
@soon  and @sabaka , please take a look.
Demand for all Checks to be thread safe - will over complicate thirdparty checks and make validation unstable. Speed is not that critical.

**Comment by rnveach** [comment]
@romani 
> Checks that are not thread safe - should be executed as it was before ..... without any thread concerns. there should be no demands for thread safety.

Correct me if I am wrong, but checks for MT, once it is done, will execute as `FileStatefulCheck` if there is no annotation which how the checks behave right now as described at 
https://github.com/checkstyle/checkstyle/blob/ebca211ef9f8a757a8973cc7e3252c80c736f9ca/src/main/java/com/puppycrawl/tools/checkstyle/FileStatefulCheck.java#L43-L44

If checks don't want to go through the hassle of making their check thread safe, which is what is required when working with multiple threads, then they shouldn't use the `GlobalStatefulCheck` annotation and stick to `FileStatefulCheck`.
`GlobalStatefulCheck` is almost the equivalent to multi-file mode (TranslationCheck) as all threads work with the same instance and retrieve all files. I don't see how you can get around this requirement and still sustain multi-threading capabilities.

**Comment by soon** [comment]
I am agree with @rnveach. However, if we decide to execute unmarked checks in the same thread this could be done by forcing each `FileSetCheck` also return a list of unmarked checks to the MT Checker, and after that MT Checker will be able to process all of them in the same thread. 

Right now MT Checker just executes a list of checks on the thread pool and does nothing besides that.

**Comment by rnveach** [comment]
> GlobalStatefulCheck annotation should be allowed ONLY on FileSetChecks, where user have ability to get call back at the very end.
`@Override public void finishProcessing() { .... }`

This makes sense as the check is multi-file validation, however there are other ways around to multi-file without `finishProcessing` that `AbstractCheck`s can do.
The example in our repository is `JavadocPackageCheck`. It uses a field to keep track of all directories processed to avoid printing the same violation multiple times. It doesn't use `finishProcessing` but is still a multi-file check.

> All FilesetChecks without annotation should be considered as GlobalStatefulCheck

I believe the implementation of the cloner in one of the PRs already does this.

> Compatibility is more in demand then speed.

The main reason people want multi-threading is speed, but they don't want to give up reliability. If they don't require speed, then multi-threading serves no purpose.

> Almost nobody demand speed.

Because the only slowdowns have been identified and issues created for them or issues were denied.

**Comment by rnveach** [comment]
> All FilesetChecks without annotation should be considered as GlobalStatefulCheck

I will have to re-confirm this case, but `GlobalStateful` shouldn't be the default annotation. It means all threads will use the same instance and so will cause run issues with multiple threads working with the same instances.

**Comment by romani** [comment]
> The example in our repository is JavadocPackageCheck

it should stay withtout annotation and be global
`public class JavadocPackageCheck extends AbstractFileSetCheck {`

> If they don't require speed, then multi-threading serves no purpose.

We have to compatible with previous versions and custom Checks.
Compatibility first, speed is second, at least for now. Some day in future we might break compatibility to improve speed, but this is should be very good speed improvement.

> It means all threads will use the same instance and so will cause run issues with multiple threads working with the same instances.

This is main point of compatibility. Ones we introduce MT, user should notice anything , even his own Checks are terrible hack, he need them. They continue to work as before. 
And main point is that we should split execution in two parts: MT-possible fileset checks and global Checks.
MT-possible fileset checks  - do any MT techniques to run.
global Checks  - just another launch on the same file set of checkstyle for global Checks.

**Comment by rnveach** [comment]
> we should split execution in two parts: MT-possible fileset checks and global Checks.

I don't agree with this.
A split of checks into 2 different runs will cause unnecessary IOs as we will have to re-read the files. If we ever had multi-file support on abstract checks, we will have to go through re-parsing too. This is also almost the same as having 2 checkstyle runs that aren't concurrent which defeats the whole purpose of multi-threading.

The current implementation of `GlobalStatefulCheck` will handle multi-file checks correctly while keeping it inline with the rest of the multi-threaded checks. There won't need to be a separation between the different types.

**Comment by romani** [comment]
> will cause unnecessary IOs as we will have to re-read the files. If we ever had multi-file support on abstract checks, we will have to go through re-parsing too.
... runs that aren't concurrent which defeats the whole purpose of multi-threading.

Second run of checkstyle execution will cost less then 1 second, but will keep compatibility.
execution from repo root:
```
$ cat config/checkstyle_checks.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
  <property name="severity" value="error"/>

  <property name="fileExtensions" value="java, properties, xml, vm, g, g4, dtd"/>

  <module name="SeverityMatchFilter">
    <!-- report all violations except ignore -->
    <property name="severity" value="ignore"/>
    <property name="acceptOnMatch" value="false"/>
  </module>
  <module name="SuppressionFilter">
    <property name="file" value="config/suppressions.xml"/>
  </module>
  <module name="SuppressWarningsFilter"/>
  <module name="SuppressWithPlainTextCommentFilter">
    <!--
      Use suppressions.xml for suppressions, this is only example.
      checkFormat will prevent suppression comments from being valid.
    -->
    <property name="checkFormat" value="IGNORETHIS"/>
    <property name="offCommentFormat" value="CSOFF\: .*"/>
    <property name="onCommentFormat" value="CSON\: .*"/>
  </module>


  <!-- Javadoc Comments -->
  <module name="JavadocPackage">
    <property name="allowLegacy" value="false"/>
  </module>

  <!-- Miscellaneous -->
  <module name="Translation">
    <property name="requiredTranslations" value="de, fr, fi, es, pt, ja, tr, zh"/>
  </module>

</module>

$ time java -jar checkstyle-8.10.1-all.jar -c config/checkstyle_checks.xml src/main/
Starting audit...
Audit done.

real	0m0.480s
user	0m0.988s
sys	0m0.120s
```

> There won't need to be a separation between the different types.

we updated only our Checks with multithreading collections to make it happen.  Take a look at PR for introduction of this annotation on TranslationCheck, is not only marker !! is it update of Check to work in our MT model.
Who will update all other Checks in the world ?

**Comment by pbludov** [comment]
>And main point is that we should split execution in two parts: MT-possible fileset checks and global Checks.
MT-possible fileset checks - do any MT techniques to run.

Can we run the parts in parallel?
All global checks in the main thread, one by one (just as before), all MT checks in additional threads.

**Comment by romani** [comment]
> Can we run the parts in parallel?

we can, but lets do this as separate PR. To do all steps by step, not by huge PR with many many complicated code changes.

**Comment by rnveach** [comment]
> Can we run the parts in parallel?

Current implementation already does this as everything will work in parallel.
Any other way would either require slowing down 1 set of threads for the other, or parsing the file twice, or keeping files in shared memory.

> Second run of checkstyle execution will cost less then 1 second

1 second for the 2nd run. With no checks the first run will be 1 second, so full run time will probably be 2 seconds. If everything was run in parallel, time would be 1 second. So splitting run into 2 is doubling the time.
It is also so low (1 second) because we have no multi-file checks. If we ever introduce them then it won't be a quick 1 second run for the 2nd.

> Who will update all other Checks in the world ?

The same people who need to update them when any other breaking change is introduced. Not all checks would need to be updated. Only multi-file modules that don't use thread-safe classes would need to be updated.

**Comment by romani** [comment]
> With no checks the first run will be 1 second, so full run time will probably be 2 seconds .. So splitting run into 2 is doubling the time.

Extending of Checkstyle execution on 1 second (as on our project) is not big deal. It is not noticeable.
If somebody run checkstyle with no checks, he just waste execution time, such user already do not care about execution time. 
this is not countable.

> The same people who need to update them when any other breaking change is introduced.

But here we have option to not do this and let them convert their logic by already using MT mode. We will force them ones we figure out what to do with our multi-file Checks. Lets not add them to main stream of Checks. Such Checks could provide extra complications in implementation. Checkstyle is single file validation tool, all other is hack and lets keep hacks on the side.
